preprocess/docenc_main.cc

```diff
@@ -0,0 +1,258 @@
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <unistd.h>
+#include <vector>
+#include "util/file_stream.hh"
+#include "util/file_piece.hh"
+#include "util/string_piece.hh"
+#include "preprocess/base64.hh"
+
+using namespace std;
+
+enum Mode {
+	COMPRESS,
+	DECOMPRESS
+};
+
+string prefix_lines(string const &input, string const prefix)
+{
+	stringstream sin(input);
+	stringstream sout;
+
+	string line;
+	while (getline(sin, line))
+		sout << prefix << line << '\n';
+
+	return sout.str();
+}
+
+size_t decode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices, bool print_document_index, bool &delimiter_encountered) {
+	size_t document_index = 0;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+
+	for (StringPiece line : in) {
+		++document_index;
+
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+			}
+		}
+
+		string document;
+		preprocess::base64_decode(line, document);
+
+		if (!delimiter_encountered && document.find(delimiter == '\n' ? string("\n\n") : string(&delimiter, 1)) != string::npos)
```

Comment:
The delimiter is a blank line?

Reply:
This was for functionality that you could put this in front of the sentence splitter, where b64filter would not work. I couldn't think of any string that would be guaranteed to get through it unscathed, no matter what sits between decode & encode. So I settled for nothing. Or a nullbyte. But that's probably more trouble for most systems.

I've only been using it for debugging, looking at documents manually. And then a blank line is quite reasonable…unless you're searching for blank lines in the input.